### DISCUSSION TITLE
ECSQL Aggregation Function?

### DISCUSSION THREAD

**Discussion by benjamin-lindsey** [iTwin/itwinjs-core#6026] [Q&A] [answered]
Hi,

Does ECSQL have a built-in aggregation function? 

Let's say we are querying for elements with a specific user label that have geometric child elements:

```sql
SELECT e.ECInstanceId AS ParentId, ge.ECInstanceId AS ChildId FROM BisCore.Element e JOIN BisCore.GeometricElement3d ge ON ge.Parent.Id=e.ECInstanceId WHERE e.UserLabel='SomeLabel' AND ge.GeometryStream IS NOT NULL
```

This will return something like this (assuming there's just a single parent):

``` JSON
[{"ParentId":"0x232","ChildId":"0x233"},{"ParentId":"0x232","ChildId":"0x234"},{"ParentId":"0x232","ChildId":"0x235"},{"ParentId":"0x232","ChildId":"0x236"}]
```

Duplicate parent ID entries. Yes, we could just create a map programmatically. But is there a way to get this grouping without further processing?

For example, in PostgreSQL we'd be able to do an `ARRAY_AG` and then `GROUP BY` the element ECInstanceId to get something like this:

``` JSON
[{"ParentId":"0x232","ChildIds":["0x233","0x234","0x235","0x236"]}]
```

Thanks!

**Comment by grigasp** [comment, marked as answer]
Hi,

ECSQL supports all SQLite's built-in functions. Have a look at [group_concat](https://www.sqlite.org/lang_aggfunc.html#group_concat) and [json_group_array](https://www.sqlite.org/json1.html#jgrouparray).

**Reply by benjamin-lindsey** [reply]
Thanks again Grigas. Great to know about the built-in functions.

```sql
SELECT e.ECInstanceId AS ParentId, JSON_GROUP_ARRAY(ge.ECInstanceId) AS ChildIds 
    FROM BisCore.Element e 
    JOIN BisCore.GeometricElement3d ge 
        ON ge.Parent.Id=e.ECInstanceId 
    WHERE e.UserLabel='SomeLabel' 
    AND ge.GeometryStream IS NOT NULL
```
Example result:
```json
[{"ParentId":"0x232","ChildIds":"[563,564,565,566]"}]
```

**Comment by khanaffan** [comment]
You can use `DISTINCT` keyword

```sql
SELECT DISTINCT e.ECInstanceId AS ParentId, ge.ECInstanceId AS ChildId FROM BisCore.Element e JOIN BisCore.GeometricElement3d ge ON ge.Parent.Id=e.ECInstanceId WHERE e.UserLabel='SomeLabel' AND ge.GeometryStream IS NOT NULL
```

Thanks.
